### PR TITLE
coord: unify logging and catalog views

### PR DESCRIPTION
**Pull request by benesch**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3959)

### CODE REVIEW

File: src/coord/src/catalog/builtin.rs [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+    };
+    pub static ref MZ_VIEW_FOREIGN_KEYS: BuiltinTable = BuiltinTable {
+        name: "mz_view_foreign_keys",
+        desc: RelationDesc::empty()
+            .with_column("child_id", ScalarType::String.nullable(false))
+            .with_column("child_column", ScalarType::Int64.nullable(false))
+            .with_column("parent_id", ScalarType::String.nullable(false))
+            .with_column("parent_column", ScalarType::Int64.nullable(false))
+            .with_column("key_group", ScalarType::Int64.nullable(false))
+            .with_key(vec![0, 1, 4]),
```
Comment (ruchirK) [inline src/coord/src/catalog/builtin.rs:199]: I know this was this way previously but this key definition jumps out as unclear - maybe a comment here about why (child_id, child_column, key_group) constitute a key would be nice? we've definitely messed this up in the past

File: src/coord/src/coord.rs [hunk 2]
```diff
@@ -2582,7 +2582,10 @@ where
 
 fn is_cat_index(id: &GlobalId) -> bool {
     // TODO(justin): do something better here?
-    CatalogView::all_views().iter().any(|v| v.index_id() == *id)
+    BUILTINS.values().any(|b| match b {
```
Comment (ruchirK) [inline src/coord/src/coord.rs:2585]: super minor nit but just want to confirm my understanding -- you could use the tables iterator here right?

File: src/coord/src/catalog/builtin.rs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+            Builtin::View(&MZ_PERF_PEEK_DURATIONS_BUCKET),
+            Builtin::View(&MZ_PERF_PEEK_DURATIONS_AGGREGATES),
+            Builtin::View(&MZ_MATERIALIZATION_FRONTIERS),
+            Builtin::View(&MZ_PERF_DEPENDENCY_FRONTIERS),
+        ];
+        builtins.into_iter().map(|b| (b.id(), b)).collect()
+    };
+}
+
+impl BUILTINS {
```
Comment (ruchirK) [inline src/coord/src/catalog/builtin.rs:444]: nice! I didn't know that lazy_static made a new type but thats pretty cool

File: src/coord/src/catalog/builtin.rs [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+            Builtin::View(&MZ_RECORDS_PER_DATAFLOW),
+            Builtin::View(&MZ_RECORDS_PER_DATAFLOW_GLOBAL),
+            Builtin::View(&MZ_PERF_ARRANGEMENT_RECORDS),
+            Builtin::View(&MZ_PERF_PEEK_DURATIONS_CORE),
+            Builtin::View(&MZ_PERF_PEEK_DURATIONS_BUCKET),
+            Builtin::View(&MZ_PERF_PEEK_DURATIONS_AGGREGATES),
+            Builtin::View(&MZ_MATERIALIZATION_FRONTIERS),
+            Builtin::View(&MZ_PERF_DEPENDENCY_FRONTIERS),
+        ];
+        builtins.into_iter().map(|b| (b.id(), b)).collect()
```
Comment (ruchirK) [inline src/coord/src/catalog/builtin.rs:440]: I think that if you duplicate an id here the map will silently drop the older value so we might want to manually do the insert here and assert that each one doesn't replace anything?

Comment (ruchirK) [reply src/coord/src/catalog/builtin.rs:440]: i think what might be nice is a separate builtin id map that we insert into that asserts that each item and index id is unique and conforms to the specified ranges? (I am also happy to do that as a follow-on to this)

**Comment by ruchirK** [comment]
This looks great btw! I just left some minor nits that you should feel free to disregard :) :)

File: src/coord/src/catalog/builtin.rs (hunk 1 above)
Comment (benesch) [reply src/coord/src/catalog/builtin.rs:199]: I'm... afraid I have no idea. I left a TODO.

File: src/coord/src/coord.rs (hunk 2 above)
Comment (benesch) [reply src/coord/src/coord.rs:2585]: Oops, yes! I added that iterator after I rewrote this method. Thanks, updated.

File: src/coord/src/catalog/builtin.rs (hunk 3 above)
Comment (benesch) [reply src/coord/src/catalog/builtin.rs:444]: Heh, yeah, comes in really handy right here!

File: src/coord/src/catalog/builtin.rs (hunk 4 above)
Comment (benesch) [reply src/coord/src/catalog/builtin.rs:440]: Yep, 100%. I actually wrote that code and then ripped it out. I'm about to land a refactor that maybe deletes the `index_id`s so I didn't want to be fighting against whatever code. I'll file an issue about sanity checking the builtins once all that lands, so we don't forget.

**Comment by benesch** [comment]
🎉🎉🎉 thanks for taking a look at this on a Sunday!

**Pull request merged**